src/python/pants/backend/python/lint/black/rules.py

```diff
@@ -87,69 +48,77 @@ def create_execute_request(
       entry_point=black.get_entry_point(),
     )
   )
+  return BlackSetup(requirements_pex=requirements_pex, config_snapshot=config_snapshot)
 
-  sources_digest = wrapped_target.target.sources.snapshot.directory_digest
 
-  merged_input_files = await Get(
-    Digest,
-    DirectoriesToMerge(
-      directories=(
-        sources_digest,
-        resolved_requirements_pex.directory_digest,
-        config_snapshot.directory_digest,
-      )
-    ),
-  )
-  return BlackSetup(config_path, resolved_requirements_pex, merged_input_files)
+@dataclass(frozen=True)
+class BlackArgs:
+  args: Tuple[str, ...]
+
+  @staticmethod
+  def create(
+    *, wrapped_target: FormattablePythonTarget, black_setup: BlackSetup, check_only: bool
+  ) -> "BlackArgs":
+    files = wrapped_target.target.sources.snapshot.files
+    pex_args = []
+    if check_only:
+      pex_args.append("--check")
+    if black_setup.config_snapshot.files:
+      pex_args.extend(["--config", black_setup.config_snapshot.files[0]])
+    # NB: For some reason, Black's --exclude option only works on recursive invocations, meaning
+    # calling Black on a directory(s) and letting it auto-discover files. However, we don't want
+    # Black to run over everything recursively under the directory of our target, as Black should
+    # only touch files in the target's `sources`. We can use `--include` to ensure that Black only
+    # operates on the files we actually care about.
+    pex_args.extend(["--include", "|".join(re.escape(f) for f in files)])
```

Comment:
Does this work with very large `files` lists? Is there some effective limit on arg size somewhere?